pkg/client/unversioned/request.go

```diff
@@ -427,8 +427,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
 		return r
 	}
 	for k, v := range params {
-		for _, vv := range v {
-			r.setParam(k, vv)
+		for _, value := range v {
+			// TODO: Move it to setParam method, once we get rid of
+			// FieldSelectorParam & LabelSelectorParam methods.
+			if k == unversioned.LabelSelectorQueryParam(r.apiVersion) && value == "" {
+				// Don't set an empty selector for backward compatibility.
```

Comment:
Yes they are orthogonal, but related in the sense of location.

There should be conversion functions for field selectors that know how to
target the field selectors.  The hard problem is that the type of the field
selector is a dynamic input, not a static input.  We may need to support
conversion options to Convert that define additional metadata that allows
the field selector conversion to target base on the input metadata
(essentially an object reference).  One scheme is not guaranteed to hold
all conversions either - so there's an inside out problem here we may have
to solve in the future.

On Wed, Dec 9, 2015 at 12:00 PM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/client/unversioned/request.go
> <https://github.com/kubernetes/kubernetes/pull/17836#discussion_r47118963>
> :
>
> > @@ -427,8 +427,36 @@ func (r *Request) VersionedParams(obj runtime.Object, convertor runtime.ObjectCo
> >  		return r
> >  	}
> >  	for k, v := range params {
> > -		for _, vv := range v {
> > -			r.setParam(k, vv)
> > +		for _, value := range v {
> > +			// TODO: Move it to setParam method, once we get rid of
> > +			// FieldSelectorParam & LabelSelectorParam methods.
> > +			if k == unversioned.LabelSelectorQueryParam(r.apiVersion) && value == "" {
> > +				// Don't set an empty selector for backward compatibility.
>
> So in other words - we should definitely clean this up, even before
> getting back to versioned ListOptions - because those are orthogonal
> problems.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17836/files#r47118963>.
>